packages/cli/test/snapshot-matcher.js

```diff
@@ -0,0 +1,205 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+'use strict';
+
+/*
+A lightweight helper for snapshot-based testing.
+- Feature inspired by https://jestjs.io/docs/en/snapshot-testing
+- Implementation inspired by node-tap
+
+This is an initial experimental version. In the (near) future, we should
+move this file to a standalone package so that all Mocha users can use it.
+*/
+
+const assert = require('assert');
+const path = require('path');
+
+module.exports = {
+  initializeSnapshots,
+};
+
+/**
+ * Create a function to match the given value against a pre-recorder snapshot.
+ *
+ * Example usage:
+ *
+ * ```js
+ * // At the top of your test file, initialize the matcher function.
+ * // You can also move this code to a shared file require()d from tests.
+ * const expectToMatchSnapshot = initializeSnapshots(
+ *   path.resolve(__dirname, '../../__snapshots__'),
+ * );
+ *
+ * // Inside your tests, call `expectToMatchSnapshot(actualValue)`
+ * describe('my feature', () => {
+ *   it('works for strings', function() {
+ *     expectToMatchSnapshot('foo\nbar');
+ *   });
+ * });
+ * ```
+ */
+function initializeSnapshots(snapshotDir) {
+  let currentTest;
+
+  beforeEach(function setupSnapshots() {
+    // eslint-disable-next-line no-invalid-this
+    currentTest = this.currentTest;
+    currentTest.__snapshotCounter = 1;
+  });
+
+  if (!process.env.UPDATE_SNAPSHOTS) {
+    return function expectToMatchSnapshot(actual) {
+      matchSnapshot(snapshotDir, currentTest, actual);
+    };
+  }
+
+  const snapshots = Object.create(null);
+  after(function updateSnapshots() {
+    for (const f in snapshots) {
```

Comment:
FIXME: This code is writing snapshot files in series using sync FS, which can be unnecessary slow. Let's rewrite it to use `async` and write tests in parallel via `Promise.all`.

Reply:
See https://github.com/strongloop/loopback-next/pull/3776